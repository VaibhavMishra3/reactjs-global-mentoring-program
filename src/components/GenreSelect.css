.genre-container {
  display: flex;
  gap: 10px;
  margin: 10px 0;
}

.genre-button {
  padding: 10px;
  border: none;
  background-color: #232323;
  color: white;
  cursor: pointer;
  border-radius: 4px;
}

.genre-button.selected {
  font-weight: bold;
  border-bottom: 2px solid #f65261;
}

.genre-button:hover {
  color: #f65261;
}
