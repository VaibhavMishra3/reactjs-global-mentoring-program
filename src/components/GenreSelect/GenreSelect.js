import React from 'react';
import { PropTypes } from 'prop-types';

class GenreSelect extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      selectedGenres: this.props.selectedGenres
    }
  }

  render() {
    return <fieldset className='d-flex'>
      {
        this.props.genres.map((genre, key) => {
          return <div key={genre} className='form-check form-check-inline'>
            <input
              data-testid='genre-select-checkbox'
              type='checkbox'
              defaultChecked={this.state.selectedGenres.includes(genre) ? 'checked' : ''}
              id={genre}
              name={genre}
              value={genre}
              onClick={this.handleClick}
              className='form-check-input'
            />
            <label
              data-testid='genre-select-label'
              role='label'
              htmlFor={genre}
              className='form-check-label'
            >
              {genre}
            </label>
          </div>
        })
      }
    </fieldset>
  }

  handleClick = (event) => {
    const selectedGenres = [...this.state.selectedGenres];
    const selectedGenre = event.target.value;

    if (selectedGenres.includes(selectedGenre)) {
      selectedGenres.splice(selectedGenres.indexOf(selectedGenre), 1);

      this.setState(prevState => ({ ...prevState, selectedGenres: selectedGenres }));
      this.props.onSelect(selectedGenres);
    } else {
      selectedGenres.push(selectedGenre);

      this.setState(prevState => ({ ...prevState, selectedGenres: selectedGenres }));
      this.props.onSelect(selectedGenres)
    }
  }

}

GenreSelect.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  selectedGenres: PropTypes.arrayOf(PropTypes.string).isRequired,
  onSelect: PropTypes.func.isRequired
};

export default GenreSelect;
