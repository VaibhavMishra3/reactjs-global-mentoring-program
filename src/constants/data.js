export const GENRES = [
  'Action',
  'Adventure',
  'Comedy',
  'Crime',
  'Fantasy',
  'History',
  'Science Fiction',
  'Thriller',
  'Western'
];

export const SORT_OPTIONS = [
  {
    name: 'Title',
    value: 'title'
  },
  {
    name: 'Release date',
    value: 'release_date'
  }
];

export const DEFAULT_SEARCH = '';
export const DEFAULT_GENRES = [];
export const DEFAULT_SORT = SORT_OPTIONS[0].value;

export const SEARCH_QUERY_PARAM = 'search';
export const GENRES_QUERY_PARAM = 'genres';
export const SORT_QUERY_PARAM = 'sort';
