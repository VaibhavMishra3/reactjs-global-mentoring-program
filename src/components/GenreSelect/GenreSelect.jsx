import React from "react";
import "./GenreSelect.css"; // Assuming you'll handle custom styling in this file

function GenreSelect({ genres, selectedGenre, onSelect }) {
  return (
    <div className="genre-container">
      {genres.map((genre) => (
        <button
          key={genre}
          className={`genre-button ${
            genre === selectedGenre ? "selected" : ""
          }`}
          onClick={() => onSelect(genre)}
        >
          {genre}
        </button>
      ))}
    </div>
  );
}

export default GenreSelect;
