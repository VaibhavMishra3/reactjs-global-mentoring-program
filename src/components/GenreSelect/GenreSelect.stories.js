import GenreSelect from './GenreSelect';
import { GENRES } from '../../constants/data.js';

export default {
  component: GenreSelect
};

export const Default = {
  args: {
    genres: GENRES,
    selectedGenres: [],
    onSelect: (genres) => console.log(`select genres: ${genres}`)
  },
  parameters: {
  }
};
