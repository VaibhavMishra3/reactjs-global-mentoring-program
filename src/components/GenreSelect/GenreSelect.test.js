import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import GenreSelect from './GenreSelect';
import { GENRES } from '../../constants/data.js';

describe('GenreSelect', () => {

  afterEach(cleanup);

  it('should match snapshot', () => {
    const genreSelect = renderer
      .create(
        <GenreSelect
          genres={GENRES}
          selectedGenres={[GENRES[1], GENRES[2]]}
          onSelect={console.log}
        />
      )
      .toJSON();

    expect(genreSelect).toMatchSnapshot();
  });

  // Test that component renders all genres passed in props
  it('should render correctly', () => {
    render(
      <GenreSelect
        genres={GENRES}
        selectedGenres={[GENRES[1], GENRES[2]]}
        onSelect={console.log}
      />
    );

    const genreLabels = screen.getAllByRole('label')
      .map(label => label.innerHTML);
    expect(genreLabels).toEqual(GENRES);
  });

  // Test that component highlights a selected genre passed in props
  it('should highlight a selected genre', () => {
    render(
      <GenreSelect
        genres={GENRES}
        selectedGenres={[GENRES[1], GENRES[2]]}
        onSelect={console.log}
      />
    );

    const genreCheckboxes = screen.getAllByRole('checkbox')
      .filter(checkbox => checkbox.checked)
      .map(checkbox => checkbox.value);
    expect(genreCheckboxes).toEqual([GENRES[1], GENRES[2]]);
  });

  // Test that after a click event on a genre button component calls "onChange" callback and passes correct genre in arguments
  it('should call the handler', () => {
    const handleSelectMock = jest.fn();

    render(
      <GenreSelect
        genres={GENRES}
        selectedGenres={[GENRES[1], GENRES[2]]}
        onSelect={handleSelectMock}
      />
    );

    const deselectingGenre = GENRES[1];
    const deselectingCheckbox = screen.getByText(deselectingGenre);
    fireEvent.click(deselectingCheckbox, { target: { innerText: deselectingCheckbox.innerHTML } });
    expect(handleSelectMock).toHaveBeenCalledWith([GENRES[2]]);

    const selectingGenre = GENRES[3];
    const selectingCheckbox = screen.getByText(selectingGenre);
    fireEvent.click(selectingCheckbox, { target: { innerText: selectingCheckbox.innerHTML } });
    expect(handleSelectMock).toHaveBeenCalledWith([GENRES[2], GENRES[3]]);
  });

});
