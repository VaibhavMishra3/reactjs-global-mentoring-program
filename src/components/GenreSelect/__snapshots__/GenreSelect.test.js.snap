// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreSelect should match snapshot 1`] = `
<fieldset
  className="d-flex"
>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="Action"
      name="Action"
      onClick={[Function]}
      type="checkbox"
      value="Action"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Action"
      role="label"
    >
      Action
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked="checked"
      id="Adventure"
      name="Adventure"
      onClick={[Function]}
      type="checkbox"
      value="Adventure"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Adventure"
      role="label"
    >
      Adventure
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked="checked"
      id="Comedy"
      name="Comedy"
      onClick={[Function]}
      type="checkbox"
      value="Comedy"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Comedy"
      role="label"
    >
      Comedy
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="Crime"
      name="Crime"
      onClick={[Function]}
      type="checkbox"
      value="Crime"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Crime"
      role="label"
    >
      Crime
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="Fantasy"
      name="Fantasy"
      onClick={[Function]}
      type="checkbox"
      value="Fantasy"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Fantasy"
      role="label"
    >
      Fantasy
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="History"
      name="History"
      onClick={[Function]}
      type="checkbox"
      value="History"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="History"
      role="label"
    >
      History
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="Science Fiction"
      name="Science Fiction"
      onClick={[Function]}
      type="checkbox"
      value="Science Fiction"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Science Fiction"
      role="label"
    >
      Science Fiction
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="Thriller"
      name="Thriller"
      onClick={[Function]}
      type="checkbox"
      value="Thriller"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Thriller"
      role="label"
    >
      Thriller
    </label>
  </div>
  <div
    className="form-check form-check-inline"
  >
    <input
      className="form-check-input"
      data-testid="genre-select-checkbox"
      defaultChecked=""
      id="Western"
      name="Western"
      onClick={[Function]}
      type="checkbox"
      value="Western"
    />
    <label
      className="form-check-label"
      data-testid="genre-select-label"
      htmlFor="Western"
      role="label"
    >
      Western
    </label>
  </div>
</fieldset>
`;
